File: python/638. Shopping Offers.py
class Solution(object):
    def shoppingOffers(self, price, special, needs):
        """
        :type price: List[int]
        :type special: List[List[int]]
        :type needs: List[int]
        :rtype: int
        """
        def shopping(special, needs):
            if not sum(needs):
                return 0
            # 过滤掉比当前needs多的礼包
            special = list(filter(lambda x: all(x[i] <= needs[i] for i in range(l)), special))
            # 如果没有可选礼包，直接返回以单品购买的价格
            if not special:
                return sum([needs[i]*price[i] for i in range(l)])
            # 回溯, 收集本次购买每种礼包的花费加上若购买该礼包后剩余needs递归的最低花费
            res = []
            for x in special:
                res.append(x[-1] + shopping(special, [needs[i] - x[i] for i in range(l)]))
            return min(res)

        l = len(needs)
        # 先过滤掉比单独购买价格高的礼包
        special = list(filter(lambda x: x[-1] < sum(x[i]*price[i] for i in range(l)), special))
        return shopping(special, needs)

    

File: go/src/leetcode/638. Shopping Offers/638. Shopping Offers.go
/**
@author ZhengHao Lou
Date    2021/10/24
*/
package main

import "fmt"

/**
https://leetcode-cn.com/problems/shopping-offers/
思路：大礼包 + 物品, 记忆化搜索

 */
const N int = 6
func shoppingOffers(price []int, special [][]int, needs []int) int {
	n := len(price)

	var cache = make(map[[N]int]int)
	var dfs func(neds [N]int) int
	dfs = func(neds [N]int) int {
		if p, ok := cache[neds]; ok {
			return p
		}
		var ans int
		// 不购买礼包时的成本
		for i := 0; i < n; i++ {
			ans += neds[i] * price[i]
		}
		// 遍历每个礼包
		for i := range special {
			next := [N]int{}
			for k := 0; k < N; k++ {
				next[k] = neds[k]
			}
			var valid = true
			for j := 0; j < n; j++ {
				if special[i][j] > neds[j] {
					valid = false
					break
				}
			}
			if !valid {
				continue
			}

			for j := 0; j < n; j++ {
				next[j] -= special[i][j]
			}
			ans = min(ans, dfs(next) + special[i][n])
		}

		cache[neds] = ans
		return ans
	}

	ned := [N]int{}
	for i := range needs {
		ned[i] = needs[i]
	}
	return dfs(ned)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(shoppingOffers([]int{2,5}, [][]int{{3,0,5},{1,2,10}}, []int{3,2}))
}